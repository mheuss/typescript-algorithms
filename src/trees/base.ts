import { ComparisonFunction } from '../comparisons/constants';
import { numericComparison } from '../comparisons/numbers';
import { stringComparison } from '../comparisons/strings';
import { ErrorCodes } from '../constants';
import { Queue } from '../queues/fifo-queue';
import { Node } from './binary-search-tree/node';
import { ITreeObject, Traversal } from './constants';

interface IGenericTreeParams<T> {
  comparisonFunction?: ComparisonFunction<T>;
  numberOfAllowedChildrenPerNode?: number;
  initializeWithDataFile?: ITreeObject<T>;
}

export class GenericTree<T> {
  protected static SelectComparisonFunction<T>(
    incomingValue: T
  ): ComparisonFunction<T> {
    if (typeof incomingValue === 'string' || incomingValue instanceof String) {
      // @ts-ignore
      return stringComparison;
    } else if (typeof incomingValue === 'number') {
      // @ts-ignore
      return numericComparison;
    } else {
      throw {
        code: ErrorCodes.COMPARISON_FUNCTION_REQUIRED,
        message:
          'A custom comparison function is needed for arrays of types other than string or numbers',
      };
    }
  }

  protected root: Node<T> | null = null;
  protected nodeCount: number = 0;
  protected comparisonFunction: ComparisonFunction<T> | undefined;
  protected maxNumberOfChildrenPerNode: number | undefined;

  constructor(options?: IGenericTreeParams<T>) {
    this.clear();
    this.comparisonFunction = options ? options.comparisonFunction : undefined;
    if (
      options?.numberOfAllowedChildrenPerNode &&
      options.numberOfAllowedChildrenPerNode > 1
    ) {
      this.maxNumberOfChildrenPerNode = options.numberOfAllowedChildrenPerNode;
    }
    if (options?.initializeWithDataFile) {
      this.loadTreeFromObject(options?.initializeWithDataFile);
    }
  }

  /**
   * Destroys the tree. Keeps the comparison function and the maxNumberOfChildren
   */
  public clear = () => {
    this.root = null;
    this.nodeCount = 0;
  };

  /**
   * Returns the total number of nodes in the search tree.
   * @returns {number}
   */
  public getNodeCount = () => {
    return this.nodeCount;
  };

  /**
   * Returns the maximum number of children each node can have.
   */
  public getMaxNumberOfChildren = () => {
    return this.maxNumberOfChildrenPerNode;
  };

  /**
   * Loads tree data into the tree
   * @param {ITreeObject} dataFile representing the tree.
   */
  public loadTreeFromObject = (dataFile: ITreeObject<T>): void => {
    if (!dataFile || !dataFile.value) {
      throw {
        code: ErrorCodes.INVALID_TREE_DATA_FILE,
        message: 'Invalid structure passed into load',
      };
    }

    this.clear();
    this.nodeCount = 1;
    this.root = this.createLoadedNode(null, dataFile);
  };

  public saveTreeFromObject = (): ITreeObject<T> | undefined => {
    if (this.root === null) {
      return undefined;
    }

    return this.createTreeAsObject(this.root);
  };

  public traverse = (order: Traversal, operation: (item: Node<T>) => void) => {
    if (this.root === null || this.root === undefined) {
      return [];
    }

    switch (order) {
      case Traversal.Breadth_First:
        return this.traversalBreadthFirst(this.root, operation);
      case Traversal.DepthFirstInOrder:
        return this.traversalDepthFirstInOrder(this.root, operation);
      case Traversal.DepthFirstPostOrder:
        return this.traversalDepthFirstPostOrder(this.root, operation);
      case Traversal.DepthFirstPreOrder:
        return this.traversalDepthFirstPreOrder(this.root, operation);
    }
  };

  protected traversalBreadthFirst(
    startingNode: Node<T>,
    operation: (item: Node<T>) => void
  ) {
    const queue = new Queue<Node<T>>();
    queue.enqueue(startingNode);

    do {
      const node = queue.dequeue();
      if (node) {
        operation(node);

        node.children.forEach(child => {
          queue.enqueue(child);
        });
      }
    } while (queue.length() !== 0);
  }

  protected traversalDepthFirstPreOrder = (
    startingNode: Node<T>,
    operation: (item: Node<T>) => void
  ) => {
    operation(startingNode);
    startingNode.children.forEach(child => {
      if (child) {
        this.traversalDepthFirstPreOrder(child, operation);
      }
    });
  };

  protected traversalDepthFirstPostOrder = (
    startingNode: Node<T>,
    operation: (item: Node<T>) => void
  ) => {
    startingNode.children.forEach(child => {
      if (child) {
        this.traversalDepthFirstPostOrder(child, operation);
      }
    });
    operation(startingNode);
  };

  protected traversalDepthFirstInOrder = (
    startingNode: Node<T>,
    operation: (item: Node<T>) => void
  ) => {
    operation(startingNode);

    this.inOrderHelper(startingNode, operation);
  };

  private inOrderHelper = (
    startingNode: Node<T>,
    operation: (item: Node<T>) => void
  ) => {
    startingNode.children.forEach(child => {
      this.inOrderHelper(child, operation);
      operation(child);
    });
  };

  private createTreeAsObject = (node: Node<T>): ITreeObject<T> => {
    const treeToReturn = {
      children: [] as Array<ITreeObject<T>>,
      value: node.value,
    };

    node.children.forEach(item => {
      treeToReturn.children.push(this.createTreeAsObject(item));
    });

    return treeToReturn;
  };

  private createLoadedNode = (
    parent: Node<T> | null,
    dataFile: ITreeObject<T>
  ) => {
    const node = new Node<T>(dataFile.value, parent);

    dataFile.children.forEach(child => {
      node.children.push(this.createLoadedNode(node, child));
      this.nodeCount++;
    });

    return node;
  };
}
