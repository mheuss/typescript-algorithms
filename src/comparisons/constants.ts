export enum Comparisons {
  LessThan = -1,
  Equal,
  GreaterThan,
}

export type ComparisonFunction<T> = (a: T, b: T) => Comparisons;
