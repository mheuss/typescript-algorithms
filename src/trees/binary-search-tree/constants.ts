export enum Sides {
  Left,
  Right,
}
