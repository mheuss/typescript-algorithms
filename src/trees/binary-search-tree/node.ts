export class Node<T> {
  public value: T;
  public children: Array<Node<T>>;
  public parent: Node<T> | null;

  constructor(value: T, parent?: Node<T> | null) {
    this.value = value;
    this.children = [];
    this.parent = parent ? parent : null;
  }
}
