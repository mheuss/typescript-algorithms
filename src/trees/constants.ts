export interface ITreeObject<T> {
  children: Array<ITreeObject<T>>;
  value: T;
}

export enum Traversal {
  Breadth_First,
  DepthFirstInOrder,
  DepthFirstPreOrder,
  DepthFirstPostOrder,
}
