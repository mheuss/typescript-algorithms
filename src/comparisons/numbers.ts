import { ErrorCodes } from '../constants';
import { Comparisons } from './constants';

/**
 * Simple numeric comparison function.
 * @param {number} a
 * @param {number} b
 * @throws {ErrorCodes.COMPARISON_ERROR} If it can't compare two numbers.
 * @returns {Comparisons}
 */
export function numericComparison(a: number, b: number): Comparisons {
  if (a > b) {
    return Comparisons.GreaterThan;
  } else if (a < b) {
    return Comparisons.LessThan;
  } else if (a === b) {
    return Comparisons.Equal;
  }

  throw {
    code: ErrorCodes.COMPARISON_ERROR,
    message: `I can not compare ${a} and ${b}`,
  };
}
