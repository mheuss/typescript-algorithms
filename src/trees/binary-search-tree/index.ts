import { ComparisonFunction, Comparisons } from "../../comparisons/constants";
import { GenericTree } from "../base";
import { Sides } from "./constants";
import { Node } from "./node";

/**
 * Binary search tree implementation.
 */
export class BinarySearchTree<T> extends GenericTree<T> {
  private static recursiveFind<T>(
    node: Node<T>,
    incomingValue: T,
    comparisonFunction: ComparisonFunction<T>
  ): Node<T> | null {
    const comparison = comparisonFunction(incomingValue, node.value);

    switch (comparison) {
      case Comparisons.GreaterThan:
        if (node.children[Sides.Right] === undefined) {
          return null;
        }
        return BinarySearchTree.recursiveFind(
          node.children[Sides.Right],
          incomingValue,
          comparisonFunction
        );
      default:
      case Comparisons.Equal:
        return node;
      case Comparisons.LessThan:
        if (node.children[Sides.Left] === undefined) {
          return null;
        }
        return BinarySearchTree.recursiveFind(
          node.children[Sides.Left],
          incomingValue,
          comparisonFunction
        );
    }
  }

  private static recursiveComparison<T>(
    node: Node<T>,
    incomingValue: T,
    comparisonFunction: ComparisonFunction<T>
  ): Node<T> | null {
    const comparison = comparisonFunction(incomingValue, node.value);

    switch (comparison) {
      case Comparisons.GreaterThan:
        if (node.children[Sides.Right] === undefined) {
          node.children[Sides.Right] = new Node(incomingValue, node);
          return node.children[Sides.Right];
        }
        return BinarySearchTree.recursiveComparison(
          node.children[Sides.Right],
          incomingValue,
          comparisonFunction
        );
      default:
      case Comparisons.Equal:
        return null;
      case Comparisons.LessThan:
        if (node.children[Sides.Left] === undefined) {
          node.children[Sides.Left] = new Node(incomingValue, node);
          return node.children[Sides.Left];
        }
        return BinarySearchTree.recursiveComparison(
          node.children[Sides.Left],
          incomingValue,
          comparisonFunction
        );
    }
  }

  constructor(comparisonFunction?: ComparisonFunction<T>) {
    super({
      comparisonFunction,
      numberOfAllowedChildrenPerNode: 2,
    });
  }

  /**
   * Will insert a node into the proper place in a tree
   * @param incomingValue
   * @returns {BinarySearchTree} Current Instance
   */
  public insert = (incomingValue: T): BinarySearchTree<T> => {
    if (this.comparisonFunction === undefined) {
      this.comparisonFunction =
        BinarySearchTree.SelectComparisonFunction<T>(incomingValue);
    }

    if (this.root === null) {
      this.root = new Node(incomingValue);
      this.nodeCount = 1;
      return this;
    }

    const returnValue = BinarySearchTree.recursiveComparison<T>(
      this.root,
      incomingValue,
      this.comparisonFunction
    );
    if (returnValue) {
      this.nodeCount++;
    }

    return this;
  };

  /**
   * Will find a node with a given value in the tree
   * @param target Value to find
   * @returns boolean
   */
  public find = (target: T): boolean => {
    return this.findNode(target) !== null;
  };

  protected findNode = (target: T): Node<T> | null => {
    if (target === null || target === undefined || this.root === null) {
      return null;
    }

    if (this.comparisonFunction === undefined) {
      this.comparisonFunction =
        BinarySearchTree.SelectComparisonFunction<T>(target);
    }

    return BinarySearchTree.recursiveFind<T>(
      this.root,
      target,
      this.comparisonFunction
    );
  };

  // @todo Implement Delete https://www.geeksforgeeks.org/binary-search-tree-set-2-delete/
}
