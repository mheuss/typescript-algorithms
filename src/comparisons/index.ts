import { ErrorCodes } from '../constants';
import { ComparisonFunction } from './constants';
import { numericComparison } from './numbers';
import { stringComparison } from './strings';

/**
 * This function is used commonly to return the correct comparison function for a type of data.
 * @param incomingArray Array of data we will be looking at
 * @param comparison Function, if provided, that'll be used for comparing.
 * @throws If there is not a comparison function provided AND we are not dealing with numbers or string
 * @constructor
 */
export function SelectComparisonFunction<T>(
  incomingArray: T[],
  comparison?: ComparisonFunction<T>
): ComparisonFunction<T> {
  const sortMechanismToBePassedIntoHelper = comparison;

  if (sortMechanismToBePassedIntoHelper === undefined) {
    if (
      typeof incomingArray[0] === 'string' ||
      incomingArray[0] instanceof String
    ) {
      // @ts-ignore
      return stringComparison;
    } else if (typeof incomingArray[0] === 'number') {
      // @ts-ignore
      return numericComparison;
    } else {
      throw {
        code: ErrorCodes.COMPARISON_FUNCTION_REQUIRED,
        message:
          'A custom comparison function is needed for arrays of types other than string or numbers',
      };
    }
  } else {
    return sortMechanismToBePassedIntoHelper;
  }
}
