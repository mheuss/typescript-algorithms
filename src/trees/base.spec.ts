import { numericComparison } from '../comparisons/numbers';
import { stringComparison } from '../comparisons/strings';
import { ErrorCodes } from '../constants';
import { GenericTree } from './base';
import { Node } from './binary-search-tree/node';
import { ITreeObject, Traversal } from './constants';

describe('Base Tree Unit tests', () => {
  it('Should return the count of nodes in a given tree', () => {
    const tree = new GenericTree<number>({ numberOfAllowedChildrenPerNode: 3 });
    expect(tree.getNodeCount()).toEqual(0);
    const NC = 'nodeCount';
    tree[NC] = 13;
    expect(tree.getNodeCount()).toEqual(13);
  });

  describe('Max kids per node', () => {
    it('Should tell us the maximum number of children each node can have', () => {
      const tree = new GenericTree<number>({
        numberOfAllowedChildrenPerNode: 3,
      });
      expect(tree.getMaxNumberOfChildren()).toEqual(3);
    });

    it('Should give us undefined, if undefined is what we pass in', () => {
      const tree = new GenericTree<number>({});
      expect(tree.getMaxNumberOfChildren()).toBeUndefined();
    });

    it('Should give us undefined, if an out of bounds number is passed in', () => {
      const tree = new GenericTree<number>({
        numberOfAllowedChildrenPerNode: -23,
      });
      expect(tree.getMaxNumberOfChildren()).toBeUndefined();
    });
  });

  describe('Load from data file in constructor', () => {
    it('Should load from data file if specific in constructor', () => {
      const tree = new GenericTree<number>({
        initializeWithDataFile: treeData,
      });
      expect(tree.getNodeCount()).toEqual(9);
    });

    it('Should throw if passed malarkey', () => {
      try {
        // tslint:disable-next-line:no-unused-expression
        new GenericTree<number>({
          initializeWithDataFile: { salad: 'yes please' } as any,
        });
        fail('Should not have got here');
      } catch (e) {
        expect(e.code).toEqual(ErrorCodes.INVALID_TREE_DATA_FILE);
      }
    });
  });

  describe('SelectComparisonFunction', () => {
    const StaticComparisonFunction = 'SelectComparisonFunction';
    const instanceComparison = 'comparisonFunction';

    it('Should select a comparison function if not selected - number', () => {
      expect(GenericTree[StaticComparisonFunction](5)).toEqual(
        numericComparison
      );
    });

    it('Should save a passed in comparator', () => {
      const tree = new GenericTree<number>({
        comparisonFunction: numericComparison,
      });
      expect(tree[instanceComparison]).toEqual(numericComparison);
    });

    it('Should select a comparison function if not selected - string', () => {
      expect(GenericTree[StaticComparisonFunction]('Whoa')).toEqual(
        stringComparison
      );
    });

    it("Should throw if it ain't got one", () => {
      try {
        GenericTree[StaticComparisonFunction]({ emf: 'Whoa' });
        fail("Shouldn't be here");
      } catch (e) {
        expect(e.code).toEqual(ErrorCodes.COMPARISON_FUNCTION_REQUIRED);
      }
    });
  });

  describe('Load/Save data into the tree', () => {
    describe('With Data', () => {
      const tree = new GenericTree<number>({
        numberOfAllowedChildrenPerNode: 3,
      });
      it('Should load data into the tree', () => {
        tree.loadTreeFromObject(treeData);
        const ROOT = 'root';
        expect(tree[ROOT]).toMatchSnapshot();
        expect(tree.getNodeCount()).toEqual(9);
      });

      it('Should export the data back out', () => {
        expect(tree.saveTreeFromObject()).toEqual(treeData);
      });
    });
    describe('Without data', () => {
      it('Should return undefined if we got nothing to save', () => {
        const tree = new GenericTree<number>({
          numberOfAllowedChildrenPerNode: 3,
        });
        expect(tree.saveTreeFromObject()).toBeUndefined();
      });
    });
  });

  describe('Traversals', () => {
    it('Should do nothing if we give it no data to work with', () => {
      const tree = new GenericTree<number>({
        comparisonFunction: numericComparison,
        numberOfAllowedChildrenPerNode: 3,
      });
      const output: number[] = [];
      const lambda = (node: Node<number>) => {
        output.push(node.value);
      };
      tree.traverse(Traversal.Breadth_First, lambda);
      expect(output).toEqual([]);
    });

    it('Should do Breadth First', () => {
      const tree = new GenericTree<number>({
        comparisonFunction: numericComparison,
        initializeWithDataFile: treeData,
        numberOfAllowedChildrenPerNode: 3,
      });
      const output: number[] = [];
      const lambda = (node: Node<number>) => {
        output.push(node.value);
      };
      tree.traverse(Traversal.Breadth_First, lambda);
      expect(output).toEqual([7, 8, 6, 5, 4, 3, 2, 1, 0]);
    });

    it('Should do depth first, preorder', () => {
      const tree = new GenericTree<number>({
        comparisonFunction: numericComparison,
        initializeWithDataFile: treeData,
        numberOfAllowedChildrenPerNode: 3,
      });
      const output: number[] = [];
      const lambda = (node: Node<number>) => {
        output.push(node.value);
      };
      tree.traverse(Traversal.DepthFirstPreOrder, lambda);
      expect(output).toEqual([7, 8, 5, 4, 3, 6, 2, 1, 0]);
    });

    it('Should do depth first, postorder', () => {
      const tree = new GenericTree<number>({
        comparisonFunction: numericComparison,
        initializeWithDataFile: treeData,
        numberOfAllowedChildrenPerNode: 3,
      });
      const output: number[] = [];
      const lambda = (node: Node<number>) => {
        output.push(node.value);
      };
      tree.traverse(Traversal.DepthFirstPostOrder, lambda);
      expect(output).toEqual([5, 4, 3, 8, 2, 1, 0, 6, 7]);
    });

    it('Should do depth first, inorder', () => {
      const tree = new GenericTree<number>({
        comparisonFunction: numericComparison,
        initializeWithDataFile: treeData,
        numberOfAllowedChildrenPerNode: 3,
      });
      const output: number[] = [];
      const lambda = (node: Node<number>) => {
        output.push(node.value);
      };
      tree.traverse(Traversal.DepthFirstInOrder, lambda);
      expect(output).toEqual([7, 5, 4, 3, 8, 2, 1, 0, 6]);
    });
  });
});

const treeData: ITreeObject<number> = {
  children: [
    {
      children: [
        {
          children: [],
          value: 5,
        },
        {
          children: [],
          value: 4,
        },
        {
          children: [],
          value: 3,
        },
      ],
      value: 8,
    },
    {
      children: [
        {
          children: [],
          value: 2,
        },
        {
          children: [],
          value: 1,
        },
        {
          children: [],
          value: 0,
        },
      ],
      value: 6,
    },
  ],
  value: 7,
};
