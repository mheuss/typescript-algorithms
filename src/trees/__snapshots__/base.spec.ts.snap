// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Base Tree Unit tests Load/Save data into the tree With Data Should load data into the tree 1`] = `
Node {
  "children": [
    Node {
      "children": [
        Node {
          "children": [],
          "parent": [Circular],
          "value": 5,
        },
        Node {
          "children": [],
          "parent": [Circular],
          "value": 4,
        },
        Node {
          "children": [],
          "parent": [Circular],
          "value": 3,
        },
      ],
      "parent": [Circular],
      "value": 8,
    },
    Node {
      "children": [
        Node {
          "children": [],
          "parent": [Circular],
          "value": 2,
        },
        Node {
          "children": [],
          "parent": [Circular],
          "value": 1,
        },
        Node {
          "children": [],
          "parent": [Circular],
          "value": 0,
        },
      ],
      "parent": [Circular],
      "value": 6,
    },
  ],
  "parent": null,
  "value": 7,
}
`;
